File: src/main/java/ge/vakho/gxt/TKEY.java
package ge.vakho.gxt;

import java.io.IOException;
import java.io.OutputStream;
import java.io.RandomAccessFile;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Objects;

import static ge.vakho.gxt.util.BufferUtils.readBytes;

public class TKEY implements Comparable<TKEY> {

    protected long tdatEntryOffset;
    protected String tdatEntryName;

    public TKEY() {
    }

    public TKEY(long tdatEntryOffset, String tdatEntryName) {
        this.tdatEntryOffset = tdatEntryOffset;
        this.tdatEntryName = tdatEntryName;
    }

    public static TKEY from(RandomAccessFile raf) throws IOException {
        TKEY tkey = new TKEY();

        // Read TDAT entry offset
        tkey.setTdatEntryOffset((long) readBytes(raf, 4, ByteBuffer::getInt));

        // Read TDAT entry name
        byte[] array = readBytes(raf, 8, ByteBuffer::array);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < array.length && array[i] != 0; i++) {
            sb.append((char) array[i]);
        }
        tkey.setTdatEntryName(sb.toString());

        return tkey;
    }

    public void writeTo(OutputStream outputStream) throws IOException {
        // Key offset (4 bytes)
        ByteBuffer bb4 = ByteBuffer.allocate(4);
        bb4.order(ByteOrder.LITTLE_ENDIAN);
        bb4.putInt((int) tdatEntryOffset);
        outputStream.write(bb4.array());

        // Key (8 bytes)
        ByteBuffer bb8 = ByteBuffer.allocate(8);
        bb8.order(ByteOrder.LITTLE_ENDIAN);
        bb8.put(tdatEntryName.getBytes());
        outputStream.write(bb8.array());
    }

    @Override
    public int compareTo(TKEY other) {
        return tdatEntryName.compareTo(other.tdatEntryName);
    }

    public long getTdatEntryOffset() {
        return tdatEntryOffset;
    }

    public void setTdatEntryOffset(long tdatEntryOffset) {
        this.tdatEntryOffset = tdatEntryOffset;
    }

    public String getTdatEntryName() {
        return tdatEntryName;
    }

    public void setTdatEntryName(String tdatEntryName) {
        this.tdatEntryName = tdatEntryName;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TKEY tkey = (TKEY) o;
        return tdatEntryOffset == tkey.tdatEntryOffset && Objects.equals(tdatEntryName, tkey.tdatEntryName);
    }

    @Override
    public int hashCode() {
        return Objects.hash(tdatEntryOffset, tdatEntryName);
    }

    @Override
    public String toString() {
        return "TKEY{" +
                "tdatEntryOffset=" + tdatEntryOffset +
                ", tdatEntryName='" + tdatEntryName + '\'' +
                '}';
    }

}

File: src/main/kotlin/ge/vakho/gxt/TKEY.kt
package ge.vakho.gxt

import java.io.IOException
import java.io.OutputStream
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * TKEY block contains [TDAT] entry offset and its name.
 *
 * TKEY block's size is always 12 (0xC) bytes.
 *
 * @see TDAT
 */
data class TKEY(var offset: UInt, var name: String) : Comparable<TKEY> {

    @Throws(IOException::class)
    fun writeTo(outputStream: OutputStream) {
        // Key offset (4 bytes)
        val bb4 = ByteBuffer.allocate(4)
        bb4.order(ByteOrder.LITTLE_ENDIAN)
        bb4.putInt(offset.toInt())
        outputStream.write(bb4.array())

        // Key (8 bytes)
        val bb8 = ByteBuffer.allocate(8)
        bb8.order(ByteOrder.LITTLE_ENDIAN)
        bb8.put(name.toByteArray())
        outputStream.write(bb8.array())
    }

    override fun compareTo(other: TKEY): Int {
        return name.compareTo(other.name)
    }

    companion object {
        @JvmStatic
        @Throws(IOException::class)
        fun from(raf: RandomAccessFile): TKEY {
            // Read offset
            val tdatEntryOffset = readBytes(raf, 4) { it.int }.toUInt()

            // Read name
            val tdatEntryName = buildString {
                val array = readBytes(raf, 8) { it.array() }
                var i = 0
                while (i < array.size && array[i] != 0.toByte()) {
                    append(array[i].toChar())
                    i++
                }
            }
            return TKEY(tdatEntryOffset, tdatEntryName);
        }
    }
}